lib/rules/hook-use-state.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * @fileoverview Ensure symmetric naming of useState hook value and setter variables
+ * @author Duncan Beevers
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+const USE_STATE_ERROR_MESSAGE = 'useStateErrorMessage';
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Ensure symmetric naming of useState hook value and setter variables',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('hook-use-state'),
+    },
+    fixable: 'code',
+    messages: {
+      [USE_STATE_ERROR_MESSAGE]: 'useState call is not destructured into value + setter pair',
+    },
+    schema: [],
+  },
+
+  create: Components.detect((context) => {
+    let reactImportLocalName;
+    let reactUseStateLocalName;
+
+    return {
+      CallExpression(node) {
+        const isReactUseStateCall = (
```

Comment:
```suggestion
        if (utils.isES5Component(node) || utils.isES6Component(node) || utils.isExplicitComponent(node) || utils.isPureComponent(node)) {
          return;
        }
        const isReactUseStateCall = (
```

this makes me want a function in `utils` that covers all the `if` things, like `utils.areHooksInvalid(node)`.

Reply:
`react-hooks/rules-of-hooks` enforces that hooks are not used in inappropriate contexts. For performance and simplicities sake, I think it would be better not to add these additional scope checks in this rule.

Reply:
We definitely can not and should not rely on that rule being enforced; performance is wildly unimportant, and i think simplicity would dictate that the rule doesn't run on places where it makes no sense for it to run.

Reply:
I stabbed around a little bit writing a test case for this, to inform and drive the implementation.
To be clear, Is the expected behavior difference from this recommendation something like this:
 * `useState` hook is used in an invalid context, this rule *does not* flag that usage
 * `useState` hook used in a valid context, this rule *does* flag that usage

```js
import React from 'react'
export default class MyClassComponent extends React.Component {
  render() {
    // CURRENTLY flagged as invalid
    //  useState call is not destructured into value + setter pair
    // EXPECTED not flagged
    //  useState call is invoked in a weird context so this rule ignores the invocation
    const result = React.useState()
  }
}
```

I can implement this kind of fencing, but the suggested utils helpers (eg; `isES5Component`, and friends) aren't quite up to the task. We'd need some other helpers (similar to `getStatelessComponent`) to establish the context.

`eslint-plugin-react-hooks` already has some plumbing for detecting this context; maybe we could leverage that work.

Ultimately though, I think the final behavior suggested here isn't really beneficial.
If we flag variable mis-namings in a bad context, so what? I don't feel that means the tool is leading users towards bad behaviors, nor do I feel that we'd actually be protecting them *better* be remaining silent in known-bad scenarios.
Ultimately, I feel the warning should be raised by `eslint-plugin-react-hooks`, and that this rule shouldn't be over-concerned with the hook's invocation context.

Reply:
It means that code in a non-react part of the codebase will get false positives; it's the reason we have so much code doing component detection.

In other words, it SHOULD be up to react and eslint-plugin-react-hooks to make sure hooks only exist in the proper place; our rule should only be warning on valid hooks.

Reply:
I don't think this will trigger false positives because it's already keyed to the names of the React imports. Unless you consider the scenario above a false positive.

Reply:
I provided a scenario above where I thought these additional checks would cause a change in this rule's behavior; in that scenario, I don't think that flagging is a bad call, even if the context is "bad" or otherwise inhospitable to hooks.

Can you help me understand a scenario where this rule might trigger a false positive, now that the React import guardrails are in place?

Reply:
Hmm, I think you might be right.

I suppose if we're certain that `useState` came from React, then it's not ever going to have a false positive, so maybe this is fine as-is. I was stuck on envisioning things for custom hooks, but that doesn't apply here.

Reply:
It is still possible to get a false positive when using `useState`, for example, when the imported `useState` is shadowed.

```ts
import { useState } from 'react'
function useState() {
  return 'whatever'
}
function MyComponent() {
  const result = useState() // flagged by hook-set-state-names
}
```

However, checking the is-a-valid-hook-context wouldn't prevent this kind of error.

Reply:
Very true. To handle that, we'll need to use the available variable scope tools to ensure that the variable isn't shadowing the import/require.

Reply:
I added handling and test cases for shadowed `useState` and `React` references.